libraries/chain/fork_database.cpp

```diff
@@ -21,79 +20,117 @@ namespace eosio::chain {
+      static void set_valid(block_state& bs, bool v) { bs.validated = v; }
+      static uint32_t last_final_block_num(const block_state& bs) { return bs.core.last_final_block_num(); }
+      static uint32_t lastest_qc_claim_block_num(const block_state& bs) { return bs.core.latest_qc_claim().block_num; }
+      static uint32_t block_height(const block_state& bs) { return bs.timestamp().slot; }
+   };
+
+   struct block_state_legacy_accessor {
+      static bool is_valid(const block_state_legacy& bs) { return bs.is_valid(); }
+      static void set_valid(block_state_legacy& bs, bool v) { bs.validated = v; }
+      static uint32_t last_final_block_num(const block_state_legacy& bs) { return bs.irreversible_blocknum(); }
+      static uint32_t lastest_qc_claim_block_num(const block_state_legacy& bs) { return bs.irreversible_blocknum(); }
+      static uint32_t block_height(const block_state_legacy& bs) { return bs.block_num(); }
+   };
+
    struct by_block_id;
-   struct by_lib_block_num;
+   struct by_best_branch;
    struct by_prev;
 
-   template<class bs>  
-   bool first_preferred( const bs& lhs, const bs& rhs ) {
-      return std::pair(lhs.irreversible_blocknum(), lhs.block_num()) > std::pair(rhs.irreversible_blocknum(), rhs.block_num());
+   template<class BS, class BSP>
+   void log_bs(const char* desc, fork_database_impl<BSP>& fork_db, const BS& lhs) {
+      using BSA = BS::fork_db_block_state_accessor;
+      dlog( "fork_db ${f}, ${d} ${bn}, last_final_block_num ${lfbn}, final_on_strong_qc_block_num ${fsbn}, lastest_qc_claim_block_num ${lbn}, block_height ${bh}, id ${id}",
+             ("f", (uint64_t)(&fork_db))("d", desc)("bn", lhs.block_num())("lfbn", BSA::last_final_block_num(lhs))("fsbn", BSA::final_on_strong_qc_block_num(lhs))("lbn", BSA::lastest_qc_claim_block_num(lhs))("bh", BSA::block_height(lhs))("id", lhs.id()) );
    }
 
-   template<class bsp>  // either [block_state_legacy_ptr, block_state_ptr], same with block_header_state_ptr
+   // match comparison of by_best_branch
+   template<class BS>
+   bool first_preferred( const BS& lhs, const BS& rhs ) {
+      using BSA = BS::fork_db_block_state_accessor;
+      return std::make_tuple(BSA::last_final_block_num(lhs), BSA::lastest_qc_claim_block_num(lhs), BSA::block_height(lhs)) >
+             std::make_tuple(BSA::last_final_block_num(rhs), BSA::lastest_qc_claim_block_num(rhs), BSA::block_height(rhs));
+   }
+
+   template<class BSP>  // either [block_state_legacy_ptr, block_state_ptr], same with block_header_state_ptr
    struct fork_database_impl {
-      using bs               = bsp::element_type;
-      using bhsp             = bs::bhsp_t;
-      using bhs              = bhsp::element_type;
-      
-      using fork_db_t        = fork_database_t<bsp>;
+      using BS               = BSP::element_type;
+      using BSAccessor       = BS::fork_db_block_state_accessor;
+      using BHSP             = BS::bhsp_t;
+      using BHS              = BHSP::element_type;
```

Comment:
these are not template parameters. I think it would be nicer and more idiomatic to have them in lowercase (like `std::vector<T>` defines `using value_type = T;`).

I'd be for somethink like this, or the types with `_t` suffix

```
      using bsp              = BSP;
      using bs               = bsp::element_type;
      using bhsp             = bs::bhsp_t;
      using bhs              = bhsp::element_type;
```

Reply:
Interesting. I'm ok with that. I do think `bsp_t` `bs_t`, etc. is better.

Reply:
Perfect, let's go for that then, I like it too!

Reply:
Maybe even for consistency add `using bsp_t = BSP;`